core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -591,6 +592,67 @@ object Validator {
 
   }
 
+  private def validateInputObject(t: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsInputType()
+    def isInputType(t: __Type): Either[__Type, Unit] = t.kind match {
+      case __TypeKind.LIST | __TypeKind.NON_NULL                         => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
+      case __TypeKind.SCALAR | __TypeKind.ENUM | __TypeKind.INPUT_OBJECT => Right(())
+      case _                                                             => Left(t)
+    }
+
+    def failValidation(msg: String, explanatoryText: String): IO[ValidationError, Unit] =
```

Comment:
If that's not too much a hassle, could you extract it and use it elsewhere in this file?